File: shared/python/course/dataset.py
"""Validation for generated eval cases (lesson 11). Pure - unit tested.

A model generating its own test data will produce duplicates, labels outside the
enum, and emails that contradict their own labels. Catch what code can catch; a
human still reviews the rest.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any

from course.triage import CATEGORIES, URGENCIES

#: An order identifier: a token with a digit, introduced by # or the word "order".
ORDER_ID_PATTERN = re.compile(r"(?:#|\border\s+#?)([A-Za-z]*-?\d[\w-]*)", re.IGNORECASE)


@dataclass(frozen=True)
class ValidationIssue:
    id: str
    problem: str


def validate_cases(
    cases: list[dict[str, Any]],
) -> tuple[list[dict[str, Any]], list[ValidationIssue]]:
    accepted: list[dict[str, Any]] = []
    issues: list[ValidationIssue] = []
    seen: set[str] = set()

    for candidate in cases:
        problems: list[str] = []
        expected = candidate.get("expected") or {}
        case_id = candidate.get("id") or ""
        text = candidate.get("input") or ""

        if not case_id:
            problems.append("missing id")
        elif case_id in seen:
            problems.append("duplicate id")

        if len(text.strip()) < 20:
            problems.append("email too short to be a realistic case")
        if expected.get("category") not in CATEGORIES:
            problems.append(f'category "{expected.get("category")}" not in the enum')
        if expected.get("urgency") not in URGENCIES:
            problems.append(f'urgency "{expected.get("urgency")}" not in the enum')

        # The label must agree with the text. This is the check that actually
        # catches bad generated data.
        looks_like_it_has_an_id = bool(ORDER_ID_PATTERN.search(text))
        if isinstance(expected.get("has_order_id"), bool) and (
            expected["has_order_id"] != looks_like_it_has_an_id
        ):
            problems.append(
                f"has_order_id={expected['has_order_id']} contradicts the email text"
            )

        if problems:
            issues.append(ValidationIssue(case_id or "(no id)", "; ".join(problems)))
            continue
        seen.add(case_id)
        accepted.append({**candidate, "reviewed": False})

    return accepted, issues

File: shared/typescript/dataset.ts
/**
 * Validation for generated eval cases (lesson 11). Pure - unit tested.
 *
 * A model generating its own test data will produce duplicates, labels outside
 * the enum, and emails that contradict their own labels. Catch what code can
 * catch; a human still reviews the rest.
 */
import { CATEGORIES, URGENCIES } from "./triage.ts";

export interface GeneratedCase {
  id: string;
  input: string;
  expected: {
    category: string;
    urgency: string;
    has_order_id: boolean;
  };
  reviewed?: boolean;
}

export interface ValidationIssue {
  id: string;
  problem: string;
}

/** An order identifier: a token with a digit, introduced by # or the word "order". */
export const ORDER_ID_PATTERN = /(?:#|\border\s+#?)([A-Za-z]*-?\d[\w-]*)/i;

export function validateCases(cases: readonly GeneratedCase[]): {
  accepted: GeneratedCase[];
  issues: ValidationIssue[];
} {
  const accepted: GeneratedCase[] = [];
  const issues: ValidationIssue[] = [];
  const seen = new Set<string>();

  for (const candidate of cases) {
    const problems: string[] = [];

    if (!candidate.id) problems.push("missing id");
    else if (seen.has(candidate.id)) problems.push("duplicate id");

    if (!candidate.input || candidate.input.trim().length < 20) {
      problems.push("email too short to be a realistic case");
    }
    if (!CATEGORIES.includes(candidate.expected?.category as never)) {
      problems.push(`category "${candidate.expected?.category}" not in the enum`);
    }
    if (!URGENCIES.includes(candidate.expected?.urgency as never)) {
      problems.push(`urgency "${candidate.expected?.urgency}" not in the enum`);
    }

    // The label must agree with the text. This is the check that actually
    // catches bad generated data.
    const looksLikeItHasAnId = ORDER_ID_PATTERN.test(candidate.input ?? "");
    if (
      typeof candidate.expected?.has_order_id === "boolean" &&
      candidate.expected.has_order_id !== looksLikeItHasAnId
    ) {
      problems.push(
        `has_order_id=${candidate.expected.has_order_id} contradicts the email text`,
      );
    }

    if (problems.length > 0) {
      issues.push({ id: candidate.id || "(no id)", problem: problems.join("; ") });
      continue;
    }
    seen.add(candidate.id);
    accepted.push({ ...candidate, reviewed: false });
  }

  return { accepted, issues };
}
